enclave-runtime/src/sidechain_impl.rs

```diff
@@ -0,0 +1,268 @@
+
+///! `SlotProposer` instance that has access to everything needed to propose a sidechain block
+pub struct SlotProposer<PB: Block, SB: SignedBlock, Pair, OcallApi, LightClient, Author> {
+	pub ocall_api: Arc<OcallApi>,
+	pub light_client: Arc<LightClient>,
+	pub author: Arc<Author>,
+	pub proposer_key: Pair,
+	pub parentchain_header: PB::Header,
+	pub shard: ShardIdentifierFor<SB>,
+	_phantom: PhantomData<PB>,
+}
+
+///! `ProposerFactory` instance containing all the data to create the `SlotProposer` for the
+/// next `Slot`
+pub struct ProposerFactory<PB: Block, Pair, OcallApi, LightClient, Author> {
+	pub ocall_api: Arc<OcallApi>,
+	pub light_client: Arc<LightClient>,
+	pub author: Arc<Author>,
+	pub pair: Pair,
+	_phantom: PhantomData<PB>,
+}
+
+impl<PB: Block, Pair, OcallApi, LightClient, Author>
+	ProposerFactory<PB, Pair, OcallApi, LightClient, Author>
+{
+	pub fn new(
+		ocall_api: Arc<OcallApi>,
+		light_client: Arc<LightClient>,
+		author: Arc<Author>,
+		pair: Pair,
+	) -> Self {
+		Self { ocall_api, light_client, author, pair, _phantom: Default::default() }
+	}
+}
+
+impl<PB: Block<Hash = H256>, SB, P, OcallApi, LightClient, Author> Environment<PB, SB>
+	for ProposerFactory<PB, P, OcallApi, LightClient, Author>
+where
+	NumberFor<PB>: BlockNumberOps,
+	SB: SignedBlock<Public = sp_core::ed25519::Public, Signature = MultiSignature> + 'static,
+	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
+	P: Pair,
+	P::Public: Encode,
+	OcallApi: EnclaveOnChainOCallApi + GetStorageVerified + 'static,
+	LightClient: Validator<PB> + Send + Sync + 'static,
+	Author: AuthorApi<H256, PB::Hash>
+		+ SendState<Hash = PB::Hash>
+		+ OnBlockCreated<Hash = PB::Hash>
+		+ Send
+		+ Sync,
+{
+	type Proposer = SlotProposer<PB, SB, P, OcallApi, LightClient, Author>;
+	type Error = ConsensusError;
+
+	fn init(
+		&mut self,
+		parent_header: &PB::Header,
+		shard: ShardIdentifierFor<SB>,
+	) -> Result<Self::Proposer, Self::Error> {
+		Ok(SlotProposer {
+			ocall_api: self.ocall_api.clone(),
+			light_client: self.light_client.clone(),
+			author: self.author.clone(),
+			proposer_key: self.pair.clone(),
+			parentchain_header: parent_header.clone(),
+			shard,
+			_phantom: PhantomData,
+		})
+	}
+}
+
+impl<PB, SB, Pair, OcallApi, LightClient, Author> Proposer<PB, SB>
+	for SlotProposer<PB, SB, Pair, OcallApi, LightClient, Author>
+where
+	PB: Block<Hash = H256>,
+	NumberFor<PB>: BlockNumberOps,
+	SB: SignedBlock<Public = sp_core::ed25519::Public, Signature = MultiSignature>,
+	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
+	OcallApi: EnclaveOnChainOCallApi,
+	LightClient: Validator<PB> + Send + Sync + 'static,
+	Author:
+		AuthorApi<H256, PB::Hash> + SendState<Hash = PB::Hash> + OnBlockCreated<Hash = PB::Hash>,
+{
+	fn propose(&self, max_duration: Duration) -> Result<Proposal<SB>, ConsensusError> {
+		let (calls, blocks) = exec_tops::<PB, SB, _, _>(
```

Comment:
ah so here we call the TOPs execution. How is it related to 'propose'?

Reply:
Propose is the very generic abstraction that substrate introduced for executing the chain's specific business logic. It's  the same here. `propose` should execute whatever is necessary to create a sidechain block afterwards. In our case, it is executing tops.

Reply:
ah I see, thanks for explaining 😄